src/PixelFlags.cc

```diff
@@ -84,60 +130,82 @@ void PixelFlagsAlgorithm::measure(
     afw::detection::Footprint const & footprint(*measRecord.getFootprint());
     func.apply(footprint);
-    if (func.getBits() & (MaskedImageT::Mask::getPlaneBitMask("EDGE") |
-                          MaskedImageT::Mask::getPlaneBitMask("NO_DATA"))) {
-        _flagHandler.setValue(measRecord, EDGE, true);
-    }
-    if (func.getBits() & MaskedImageT::Mask::getPlaneBitMask("BAD")) {
-        _flagHandler.setValue(measRecord, BAD, true);
-    }
-    if (func.getBits() & MaskedImageT::Mask::getPlaneBitMask("INTRP")) {
-        _flagHandler.setValue(measRecord, INTERPOLATED, true);
-    }
-    if (func.getBits() & MaskedImageT::Mask::getPlaneBitMask("SAT")) {
-        _flagHandler.setValue(measRecord, SATURATED, true);
-    }
-    if (func.getBits() & MaskedImageT::Mask::getPlaneBitMask("CR")) {
-        _flagHandler.setValue(measRecord, CR, true);
+
+    // Set the EDGE flag if the bitmask has NO_DATA set
+    try {
+        if (func.getBits() & MaskedImageT::Mask::getPlaneBitMask("NO_DATA")){
+            measRecord.set(_anyKeys.at("EDGE"),true);
+        }
+    } catch (pex::exceptions::InvalidParameterError & err){
+        throw LSST_EXCEPT(FatalAlgorithmError, err.what());
     }
 
+    // update the source record for the any keys
+    updateFlags(_anyKeys, func, measRecord);
+
     // Check for bits set in the 3x3 box around the center
-    afw::geom::Point2I llc(afw::image::positionToIndex(center.getX())-1, afw::image::positionToIndex(center.getY()) - 1);
+    afw::geom::Point2I llc(afw::image::positionToIndex(center.getX()) - 1,
+                           afw::image::positionToIndex(center.getY()) - 1);
 
     afw::detection::Footprint const middle(afw::geom::BoxI(llc, afw::geom::ExtentI(3))); // central 3x3
     func.apply(middle);
-    if (func.getBits() & MaskedImageT::Mask::getPlaneBitMask("INTRP")) {
-        _flagHandler.setValue(measRecord, INTERPOLATED_CENTER, true);
-    }
-    if (func.getBits() & MaskedImageT::Mask::getPlaneBitMask("SAT")) {
-        _flagHandler.setValue(measRecord, SATURATED_CENTER, true);
-    }
-    if (func.getBits() & MaskedImageT::Mask::getPlaneBitMask("CR")) {
-        _flagHandler.setValue(measRecord, CR_CENTER, true);
-    }
-    _flagHandler.setValue(measRecord, FAILURE, false);
+
+    // Update the flags which have to do with the center of the footprint
+    updateFlags(_centerKeys, func, measRecord);
+
 }
 
 void PixelFlagsAlgorithm::fail(afw::table::SourceRecord & measRecord, MeasurementError * error) const {
-    _flagHandler.handleFailure(measRecord, error);
+    measRecord.set(_generalFailure, true);
+    // Because this algorithm does not explicitly throw a MeasurementError we
+    // expect this function to never be executed, and hence error would be
+    // null. If it is run there is a very unexpected behavior and the assert
+    // statement will immediately indicated to the user that something
+    // unexpected has gone wrong 
+    assert (error == NULL);
```

Comment:
I'm not sure the above is true, and it's certainly confusing!

First, my understanding is that `fail()` is called for every unhandled exception in the plugin, not just a `MeasurementError`. The fact you don't expect to throw a `MeasurementError` doesn't mean this won't be called.

Second, what does it mean to assert (either in the comment or in the code) that something is `NULL` if it's never going to exist because the function isn't going to be called?

I think you are asserting that any exceptions which are thrown are not `MeasurementError`s. I'm still not really sure why, and the above comment didn't help I'm afraid!